In [1]:
import os
import pandas as pd

Read in the data

In [77]:
data = pd.read_excel('/Users/elizabeth 1/Library/CloudStorage/Box-Box/Wu Lab/Project - 2024/Functional screening/240318_Natalie_Moe requested genes Query Sigma 226 for 162.xlsx')
data = data.loc[:,['Quadrant/DNA plate #', 'Row', 'Col', 'Moe requested genes_Name']]

In [80]:
data = data.rename(columns={'Quadrant/DNA plate #' : 'plate', 'Moe requested genes_Name' : 'gene'})
data

plate Row  Col     gene
0      1.0   C    3    Ndrg2
1      1.0   H    4   Ube2l6
2      1.0   D    9     Dlx1
3      1.0   H    1    Radil
4      1.0   D    7     sord
..     ...  ..  ...      ...
221    NaN   D   12  poldip3
222    NaN   A   10     Prph
223    NaN   B   12      Ret
224    NaN   C    8  Slc32a1
225    NaN   A    4     tfrc

[226 rows x 4 columns]

In [74]:
print(len(data))
# Use a Boolean condition directly to filter out rows where 'A' is not equal to 2
data = data[(data['plate'] == 1.0) | (data['plate'] == 2.0)]       
print(len(data))     

226
192


Create the Plate Map Dataframes

In [75]:
cols = list(range(1,25)) # half-open range (start @ start, up to but not including stop) -- also used for list slicing & iteration/loops
rows = [chr(i) for i in range(ord('A'), ord('H')+1)] # list comprehension
# ord for ordinal --> numeric value assigned to a character, returns char's ASCII val. Go up to H+1 b/c of half-open range (any string char has an ASCII val)
# chr(i) returns character that corresponds to the number, the numbers 'i' being the ASCII values from A-I

In [76]:
# create empty dataframes
plate_1 = pd.DataFrame(index=rows, columns=cols)
plate_2 = pd.DataFrame(index=rows, columns=cols) # can't se it equal to plate_1 bc plate_2 will be a variable pointing to plate_1, so any changes in plate_2 will be reflected in plate_1 and vice versa
plate_1.index.name = 'Plate 1'
plate_2.index.name = 'Plate 2'
plate_1

1    2    3    4    5    6    7    8    9    10  ...   15   16   17  \
Plate 1                                                    ...                  
A        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
B        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
C        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
D        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
E        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
F        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
G        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
H        NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   

          18   19   20   21   22   23   24  
Plate 1                                     
A        NaN  NaN  NaN  NaN  NaN  NaN  NaN  
B        NaN  NaN  NaN  NaN  NaN  NaN  NaN  
C        NaN  NaN  NaN  NaN  NaN  NaN  NaN  
D        NaN  NaN  NaN  NaN  NaN  NaN  NaN  
E        NaN  NaN  NaN  NaN  NaN  NaN  NaN  
F        NaN  NaN  NaN  NaN  NaN  NaN  NaN  
G        NaN  NaN  NaN  NaN  NaN  NaN  NaN  
H        NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[8 rows x 24 columns]

In [82]:
data['Col'].dtype

dtype('int64')

In [83]:
for index, row in data.iterrows():
    plate = row['plate']
    gene = row['gene']   
    plate_row = row['Row'] 
    plate_col = row['Col']
    
    if plate == 1.0:
        plate_1.loc[plate_row,plate_col] = gene # use loc instead of iloc to access by index name instead of #
    elif plate == 2.0:
        plate_2.loc[plate_row,plate_col] = gene
        
plate_1
        

1          2         3       4       5       6       7   \
Plate 1                                                                  
A             NaN  Rab11fip4       NaN  Fbxo17     NaN    Bcam     NaN   
B        Slc9a3r2        NaN   Lurap1l     NaN    Osr2    Rpia     NaN   
C         Ccdc85c       Tcf7  Calcoco2   Tubg2     Ret  Gpr157  Higd1a   
D             NaN    B4galt3       NaN  Ptpn18     NaN     NaN   Enpep   
E             NaN        NaN     Esrrb   Tead2  Igfbp2    Igf2   Batf3   
F          Slc2a3     Higd1a      Lhx6     NaN    Cib2    myo6   Mypop   
G             Sp6        NaN      Glo1    Wnt4     NaN    Npr2    Npr3   
H            Npr2       Npr2       NaN  Ube2l6     Fes     NaN  Higd1a   

              8        9       10  ...   15   16   17   18   19   20   21  \
Plate 1                            ...                                      
A        Slc29a4      NaN     NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
B           Mdfi  Spata2l   Amhr2  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
C            NaN      NaN     NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
D         Eps8l2  Txndc16    Aig1  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
E            NaN   Gpr157    Gfer  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
F          Nxph4  Smarcd1   Gsta3  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
G          Foxa3   Sema4d  Tex101  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
H            NaN     Lcn2     vim  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

          22   23   24  
Plate 1                 
A        NaN  NaN  NaN  
B        NaN  NaN  NaN  
C        NaN  NaN  NaN  
D        NaN  NaN  NaN  
E        NaN  NaN  NaN  
F        NaN  NaN  NaN  
G        NaN  NaN  NaN  
H        NaN  NaN  NaN  

[8 rows x 24 columns]

In [84]:
# save the unlinked plate maps
os.chdir('/Users/elizabeth 1/Library/CloudStorage/Box-Box/Wu Lab/Project - statin/8. RNA-seq/Elizabeth/LW15 analysis/LW15-Functional-Screening/plate_map')
plate_1.to_html('plate_1.html')
plate_2.to_html('plate_2.html')

### Troubleshooting
check how many duplicate well instances there are

In [85]:
data

plate Row  Col     gene
0      1.0   C    3    Ndrg2
1      1.0   H    4   Ube2l6
2      1.0   D    9     Dlx1
3      1.0   H    1    Radil
4      1.0   D    7     sord
..     ...  ..  ...      ...
221    NaN   D   12  poldip3
222    NaN   A   10     Prph
223    NaN   B   12      Ret
224    NaN   C    8  Slc32a1
225    NaN   A    4     tfrc

[226 rows x 4 columns]

In [86]:
data_1 = data[data['plate'] == 1.0] 
data_2 = data[data['plate'] == 2.0]       

note:
- if there were to be a full plate map, there's 8*24 = 192 wells per plate. 
- but the data var with plates 1&2 has 226 rows (should have 384 if complete)
- and each plate 1 and plate 2 have 96 rows, respectively

In [91]:
wells = []
for index, row in data_1.iterrows():
    row_name = row['Row']
    col_name = str(row['Col'])
    well = row_name + col_name
    wells.append(well)
    

counts_dict = {}

for well in wells:
    if well not in counts_dict:
        counts_dict[well] = 1
    else:
        counts_dict[well] += 1

for well, num in counts_dict.items():
    if num >1:
        print(well, num)

C3 2
D9 3
H1 2
D7 2
H11 2
D11 2
B5 2
A11 3
D12 2
C5 3
A2 2
G7 5
F6 2
E11 2
H7 2
D4 2
E9 2
G1 2
C2 2
F8 2
F1 2
E12 2
B8 2
B6 2
F2 2


In [118]:
def duplicate_wells(plate):
    wells = {}
    for index, row in plate.iterrows():
        # print(row)
        row_name = row['Row']
        col_name = str(row['Col'])
        well = row_name + col_name
        wells[index] = well
    
    counts_dict = {} # {A1:
    
    for index, well in wells.items():
        if well in counts_dict:
            counts_dict[well].append(index)
        else:
            counts_dict[well] = []
            counts_dict[well].append(index)
    
    for well, indices in counts_dict.items():
       indices = [index + 2 for index in indices] # update by 2 to correct for the indexing
       counts_dict[well] = indices
    
    for well, indices in counts_dict.items():
        if len(indices) > 1:
            print(well, indices)
            
duplicate_wells(data_2)


F10 [101, 106, 146, 183]
D2 [102, 131, 136, 141, 163]
F4 [103, 126, 151, 158]
G11 [104, 119]
E3 [105, 169]
B6 [108, 153]
F6 [109, 129]
H5 [111, 114]
A9 [112, 145]
D3 [115, 157, 166]
D9 [116, 188]
C7 [117, 124, 134, 154]
B12 [123, 133]
D6 [125, 127]
H4 [130, 132, 182]
H9 [139, 143]
E4 [142, 191]
A8 [144, 162, 172]
A6 [152, 161]
A1 [171, 190]
E2 [173, 181]
D7 [178, 187]


In [119]:
duplicate_wells(data_1)

C3 [2, 93]
D9 [4, 61, 74]
H1 [5, 20]
D7 [6, 83]
H11 [7, 63]
D11 [8, 71]
B5 [9, 35]
A11 [10, 58, 65]
D12 [11, 80]
C5 [13, 32, 37]
A2 [15, 89]
G7 [18, 33, 60, 76, 88]
F6 [22, 95]
E11 [27, 86]
H7 [34, 72]
D4 [38, 73]
E9 [39, 56]
G1 [41, 82]
C2 [42, 52]
F8 [50, 59]
F1 [54, 64]
E12 [57, 85]
B8 [62, 94]
B6 [70, 84]
F2 [96, 97]


In [110]:
list = [1,2,3,4]
[item + 5 for item in list]

[6, 7, 8, 9]